packages/@react-spectrum/s2/src/TreeView.tsx

```diff
@@ -0,0 +1,497 @@
+import {IconContext} from './Icon';
+import {isAndroid} from '@react-aria/utils';
+import {raw} from '../style/style-macro' with {type: 'macro'};
+import React, {createContext, forwardRef, isValidElement, JSXElementConstructor, ReactElement, useContext, useMemo, useRef} from 'react';
+import {Text, TextContext} from './Content';
+import {useButton} from '@react-aria/button';
+import {useDOMRef} from '@react-spectrum/utils';
+import {useLocale} from '@react-aria/i18n';
+
+interface S2TreeProps {
+  // Only detatched is supported right now with the current styles from Spectrum
+  isDetached?: boolean,
+  onAction?: (key: Key) => void,
+  // not fully supported yet
+  isEmphasized?: boolean
+}
+
+export interface TreeViewProps extends Omit<RACTreeProps<any>, 'style' | 'className' | 'onRowAction' | 'selectionBehavior' | 'onScroll' | 'onCellAction' | 'dragAndDropHooks'>, UnsafeStyles, S2TreeProps {
+  /** Spectrum-defined styles, returned by the `style()` macro. */
+  styles?: StylesPropWithHeight
+}
+
+export interface TreeViewItemProps<T extends object = object> extends Omit<RACTreeItemProps, 'className' | 'style'> {
+  /** Whether this item has children, even if not loaded yet. */
+  hasChildItems?: boolean,
+  /** A list of child tree item objects used when dynamically rendering the tree item children. */
+  childItems?: Iterable<T>
+}
+
+interface TreeRendererContextValue {
+  renderer?: (item) => ReactElement<any, string | JSXElementConstructor<any>>
+}
+const TreeRendererContext = createContext<TreeRendererContextValue>({});
+
+function useTreeRendererContext(): TreeRendererContextValue {
+  return useContext(TreeRendererContext)!;
+}
+
+
+let InternalTreeContext = createContext<{isDetached?: boolean, isEmphasized?: boolean}>({});
+
+// TODO: the below is needed so the borders of the top and bottom row isn't cut off if the TreeView is wrapped within a container by always reserving the 2px needed for the
+// keyboard focus ring. Perhaps find a different way of rendering the outlines since the top of the item doesn't
+// scroll into view due to how the ring is offset. Alternatively, have the tree render the top/bottom outline like it does in Listview
+const tree = style({
+  userSelect: 'none',
+  minHeight: 0,
+  minWidth: 0,
+  width: 'full',
+  overflow: 'auto',
+  boxSizing: 'border-box',
+  justifyContent: {
+    isEmpty: 'center'
+  },
+  alignItems: {
+    isEmpty: 'center'
+  },
+  height: {
+    isEmpty: 'full'
+  },
+  '--indent': {
+    type: 'width',
+    value: 16
+  }
+}, getAllowedOverrides({height: true}));
+
+function TreeView(props: TreeViewProps, ref: DOMRef<HTMLDivElement>) {
+  let {children, isDetached, isEmphasized} = props;
+
+  let renderer;
+  if (typeof children === 'function') {
+    renderer = children;
+  }
+
+  let domRef = useDOMRef(ref);
+
+  let layout = useMemo(() => {
+    return new UNSTABLE_ListLayout({
+      rowHeight: isDetached ? 44 : 40
```

Comment:
I'm seeing conflicting designs in Figma for the spacing between the detached rows, but the tokens file says that it should use `tree-view-item-to-item-detached` which is 2px whereas the design file has 8px between each row...

Reply:
looks like it should be 2, i've moved it back to that. I'll raise with design after I have a build

Reply:
looks like it should be 2, i've moved it back to that. I'll raise with design after I have a build